lib/spack/spack/spec.py

```diff
@@ -1649,11 +1649,24 @@ def node_dict_with_hashes(self, hash=ht.dag_hash):
         spec is concrete, the full hash is added as well.  If 'build' is in
         the hash_type, the build hash is also added. """
         node = self.to_node_dict(hash)
-        node[self.name]['hash'] = self.dag_hash()
+
         if self.concrete:
-            node[self.name]['full_hash'] = self.full_hash()
-        if 'build' in hash.deptype:
-            node[self.name]['build_hash'] = self.build_hash()
+            # if the spec is concrete, all hashes have already been
+            # computed, DAG hash is required, so always add it
+            node[self.name]['hash'] = self.dag_hash()
+
+            # full_hash and build_hash may not have been computed by
+            # older spack versions, so they may not be on a concrete spec
+            # read in from somewhere. We want to avoid computing
+            # full_hash or build_hash on old specs, as we likely don't
+            # have the information needed to do it right (packages,
+            # patches, etc. have likely changed). So we do not attempt to
+            # recompute full_hash or build_hash for old specs, and we
+            # only write these if they're present.
+            if self._full_hash:
+                node[self.name]['full_hash'] = self._full_hash
```

Comment:
Sorry, I intended my [comment](https://github.com/spack/spack/pull/19672#pullrequestreview-521844640) to go here.